File: C++/06. 从尾到头打印链表.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

// 方法一：辅助栈
class Solution {
   public:
    vector<int> reversePrint(ListNode* head) {
        stack<int> stk;
        ListNode* p = head;
        while (p) {
            stk.push(p->val);
            p = p->next;
        }
        vector<int> res;
        while (!stk.empty()) {
            res.push_back(stk.top());
            stk.pop();
        }
        return res;
    }
};

// 方法二：递归
class Solution {
   public:
    vector<int> res;
    vector<int> reversePrint(ListNode* head) {
        if (head) {
            reversePrint(head->next);
            res.push_back(head->val);
        }
        return res;
    }
};

File: Java/06. 从尾到头打印链表.java
class Solution {
    public int[] reversePrint(ListNode head) {
        Stack<Integer> stk = new Stack<Integer>();
        ListNode p = head;
        int size = 0;
        while (p != null) {
            stk.push(p.val);
            size++;
            p = p.next;
        }
        int[] res = new int[size];
        for (int i = 0; i < size; i++) {
            res[i] = stk.pop();
        }
        return res;
    }
}

class Solution {

    private ArrayList<Integer> res = new ArrayList<>();

    public int[] reversePrint(ListNode head) {
        if (head != null) {
            reversePrint(head.next);
            res.add(head.val);
        }
        return res.stream().mapToInt(i -> i).toArray();
    }
}